File: tests/cross-bench/bench-richards.py
# richards: OS kernel scheduling simulation. Five tasks (idler, worker,
# two handlers, two devices) driven by a packet-passing scheduler.
# Stresses allocation, linked-list mutation, and indirect dispatch.
# A correct run produces queueCount = 232625 and holdCount = 93050.
#
# Ported from the Are-We-Fast-Yet Python benchmark (richards.py); harness
# stripped, idle-task count raised to 100000 to match bench-richards.lua.

_NO_TASK = None
_NO_WORK = None

_IDLER = 0
_WORKER = 1
_HANDLER_A = 2
_HANDLER_B = 3
_DEVICE_A = 4
_DEVICE_B = 5

_NUM_TYPES = 6

_DEVICE_PACKET_KIND = 0
_WORK_PACKET_KIND = 1

_DATA_SIZE = 4

_COUNT = 100000
_EXPECTED_QUEUE_COUNT = 232625
_EXPECTED_HOLD_COUNT = 93050


class _RBObject:
    @staticmethod
    def append(packet, queue_head):
        packet.link = _NO_WORK
        if _NO_WORK == queue_head:
            return packet

        mouse = queue_head

        while True:
            link = mouse.link
            if link == _NO_WORK:
                break
            mouse = link

        mouse.link = packet
        return queue_head


class _Scheduler(_RBObject):
    def __init__(self):
        self._task_list = _NO_TASK
        self._current_task = _NO_TASK
        self._current_task_identity = 0

        self._task_table = [_NO_TASK] * _NUM_TYPES

        self._queue_count = 0
        self._hold_count = 0

    def create_device(self, identity, priority, work, state):
        data = _DeviceTaskDataRecord()

        def fn(function_work, data_record):
            if _NO_WORK == function_work:
                function_work = data_record.pending
                if _NO_WORK == function_work:
                    return self.mark_waiting()

                data_record.pending = _NO_WORK
                return self.queue_packet(function_work)

            data_record.pending = function_work
            return self._hold_self()

        self.create_task(identity, priority, work, state, data, fn)

    def create_handler(self, identity, priority, work, state):
        data = _HandlerTaskDataRecord()

        def fn(work_arg, data_record):
            if _NO_WORK != work_arg:
                if _WORK_PACKET_KIND == work_arg.kind:
                    data_record.work_in_add(work_arg)
                else:
                    data_record.device_in_add(work_arg)

            work_packet = data_record.work_in
            if _NO_WORK == work_packet:
                return self.mark_waiting()

            count = work_packet.datum
            if count >= _DATA_SIZE:
                data_record.work_in = work_packet.link
                return self.queue_packet(work_packet)

            device_packet = data_record.device_in
            if _NO_WORK == device_packet:
                return self.mark_waiting()

            data_record.device_in = device_packet.link
            device_packet.datum = work_packet.data[count]
            work_packet.datum = count + 1
            return self.queue_packet(device_packet)

        self.create_task(identity, priority, work, state, data, fn)

    def create_idler(self, identity, priority, work, state):
        data = _IdleTaskDataRecord()

        def fn(_work, data_record):
            data_record.count = data_record.count - 1
            if 0 == data_record.count:
                return self._hold_self()

            if 0 == (data_record.control & 1):
                data_record.control = data_record.control // 2
                return self.release(_DEVICE_A)

            data_record.control = (data_record.control // 2) ^ 53256
            return self.release(_DEVICE_B)

        self.create_task(identity, priority, work, state, data, fn)

    def create_packet(self, link, identity, kind):
        return _Packet(link, identity, kind)

    def create_task(self, identity, priority, work, state, data, fn):
        t = _TaskControlBlock(
            self._task_list, identity, priority, work, state, data, fn
        )
        self._task_list = t
        self._task_table[identity] = t

    def create_worker(self, identity, priority, work, state):
        data = _WorkerTaskDataRecord()

        def fn(work, data_record):
            if _NO_WORK == work:
                return self.mark_waiting()

            data_record.destination = (
                _HANDLER_B if _HANDLER_A == data_record.destination else _HANDLER_A
            )

            work.identity = data_record.destination
            work.datum = 0
            for i in range(_DATA_SIZE):
                data_record.count = data_record.count + 1
                if data_record.count > 26:
                    data_record.count = 1
                work.data[i] = 65 + data_record.count - 1

            return self.queue_packet(work)

        self.create_task(identity, priority, work, state, data, fn)

    def start(self):
        self.create_idler(_IDLER, 0, _NO_WORK, _TaskState.with_running())
        wkq = self.create_packet(_NO_WORK, _WORKER, _WORK_PACKET_KIND)
        wkq = self.create_packet(wkq, _WORKER, _WORK_PACKET_KIND)

        self.create_worker(_WORKER, 1000, wkq, _TaskState.with_waiting_with_packet())
        wkq = self.create_packet(_NO_WORK, _DEVICE_A, _DEVICE_PACKET_KIND)
        wkq = self.create_packet(wkq, _DEVICE_A, _DEVICE_PACKET_KIND)
        wkq = self.create_packet(wkq, _DEVICE_A, _DEVICE_PACKET_KIND)

        self.create_handler(
            _HANDLER_A, 2000, wkq, _TaskState.with_waiting_with_packet()
        )
        wkq = self.create_packet(_NO_WORK, _DEVICE_B, _DEVICE_PACKET_KIND)
        wkq = self.create_packet(wkq, _DEVICE_B, _DEVICE_PACKET_KIND)
        wkq = self.create_packet(wkq, _DEVICE_B, _DEVICE_PACKET_KIND)

        self.create_handler(
            _HANDLER_B, 3000, wkq, _TaskState.with_waiting_with_packet()
        )
        self.create_device(_DEVICE_A, 4000, _NO_WORK, _TaskState.with_waiting())
        self.create_device(_DEVICE_B, 5000, _NO_WORK, _TaskState.with_waiting())

        self.schedule()

        return self._queue_count, self._hold_count

    def find_task(self, identity):
        t = self._task_table[identity]
        if _NO_TASK == t:
            raise Exception("find_task failed")
        return t

    def _hold_self(self):
        self._hold_count += 1
        self._current_task.set_task_holding(True)
        return self._current_task.link

    def queue_packet(self, packet):
        task = self.find_task(packet.identity)
        if _NO_TASK == task:
            return _NO_TASK

        self._queue_count += 1

        packet.link = _NO_WORK
        packet.identity = self._current_task_identity
        return task.add_input_and_check_priority(packet, self._current_task)

    def release(self, identity):
        task = self.find_task(identity)
        if _NO_TASK == task:
            return _NO_TASK

        task.set_task_holding(False)

        if task.priority > self._current_task.priority:
            return task
        return self._current_task

    def mark_waiting(self):
        self._current_task.set_task_waiting(True)
        return self._current_task

    def schedule(self):
        self._current_task = self._task_list
        while _NO_TASK is not self._current_task:
            if self._current_task.is_task_holding_or_waiting():
                self._current_task = self._current_task.link
            else:
                self._current_task_identity = self._current_task.identity
                self._current_task = self._current_task.run_task()


class _DeviceTaskDataRecord(_RBObject):
    def __init__(self):
        self.pending = _NO_WORK


class _HandlerTaskDataRecord(_RBObject):
    def __init__(self):
        self.work_in = _NO_WORK
        self.device_in = _NO_WORK

    def device_in_add(self, packet):
        self.device_in = self.append(packet, self.device_in)

    def work_in_add(self, packet):
        self.work_in = self.append(packet, self.work_in)


class _IdleTaskDataRecord(_RBObject):
    def __init__(self):
        self.control = 1
        self.count = _COUNT


class _Packet(_RBObject):
    def __init__(self, link, identity, kind):
        self.link = link
        self.kind = kind
        self.identity = identity
        self.datum = 0
        self.data = [0] * _DATA_SIZE


class _TaskState(_RBObject):
    def __init__(self):
        self._task_holding = False
        self._task_waiting = False
        self._packet_pending = False

    def is_packet_pending(self):
        return self._packet_pending

    def is_task_waiting(self):
        return self._task_waiting

    def is_task_holding(self):
        return self._task_holding

    def set_task_holding(self, task):
        self._task_holding = task

    def set_task_waiting(self, task):
        self._task_waiting = task

    def packet_pending(self):
        self._packet_pending = True
        self._task_waiting = False
        self._task_holding = False
        return self

    def running(self):
        self._packet_pending = self._task_waiting = self._task_holding = False
        return self

    def waiting(self):
        self._packet_pending = self._task_holding = False
        self._task_waiting = True
        return self

    def waiting_with_packet(self):
        self._task_holding = False
        self._task_waiting = self._packet_pending = True
        return self

    def is_task_holding_or_waiting(self):
        return self._task_holding or (not self._packet_pending and self._task_waiting)

    def is_waiting_with_packet(self):
        return self._packet_pending and self._task_waiting and not self._task_holding

    @classmethod
    def with_running(cls):
        return cls().running()

    @classmethod
    def with_waiting(cls):
        return cls().waiting()

    @classmethod
    def with_waiting_with_packet(cls):
        return cls().waiting_with_packet()


class _TaskControlBlock(_TaskState):
    def __init__(
        self,
        link,
        identity,
        priority,
        initial_work_queue,
        initial_state,
        private_data,
        fn,
    ):
        super().__init__()
        self.link = link
        self.identity = identity
        self.function = fn
        self.priority = priority
        self._input = initial_work_queue
        self._handle = private_data

        self._packet_pending = initial_state.is_packet_pending()
        self._task_waiting = initial_state.is_task_waiting()
        self._task_holding = initial_state.is_task_holding()

    def add_input_and_check_priority(self, packet, old_task):
        if _NO_WORK == self._input:
            self._input = packet
            self._packet_pending = True
            if self.priority > old_task.priority:
                return self
        else:
            self._input = self.append(packet, self._input)

        return old_task

    def run_task(self):
        if self.is_waiting_with_packet():
            message = self._input
            self._input = message.link
            if _NO_WORK == self._input:
                self.running()
            else:
                self.packet_pending()
        else:
            message = _NO_WORK

        return self.function(message, self._handle)


class _WorkerTaskDataRecord(_RBObject):
    def __init__(self):
        self.destination = _HANDLER_A
        self.count = 0


queue_count, hold_count = _Scheduler().start()
if queue_count != _EXPECTED_QUEUE_COUNT or hold_count != _EXPECTED_HOLD_COUNT:
    raise SystemExit(f"bad: queueCount={queue_count} holdCount={hold_count}")
print(queue_count + hold_count)

File: tests/cross-bench/bench-richards.lua
-- richards: OS kernel scheduling simulation. Five tasks (idler, worker,
-- two handlers, two devices) driven by a packet-passing scheduler.
-- Stresses allocation, linked-list mutation, and indirect dispatch.
-- A correct run produces queueCount = 2322 and holdCount = 928.

local ID_IDLE       = 0
local ID_WORKER     = 1
local ID_HANDLER_A  = 2
local ID_HANDLER_B  = 3
local ID_DEVICE_A   = 4
local ID_DEVICE_B   = 5
local NUMBER_OF_IDS = 6

local KIND_DEVICE = 0
local KIND_WORK   = 1

local DATA_SIZE = 4
local COUNT     = 100000

local band, bor, bnot, rshift, bxor
do
  local ok, bitlib = pcall(require, "bit")   -- LuaJIT / LuaBitOp
  if ok then
    band, bor, bnot, rshift, bxor =
      bitlib.band, bitlib.bor, bitlib.bnot, bitlib.rshift, bitlib.bxor
  else                                        -- Lua 5.3+: native operators
    band   = load("return function(a, b) return a & b end")()
    bor    = load("return function(a, b) return a | b end")()
    bnot   = load("return function(a) return ~a end")()
    rshift = load("return function(a, n) return a >> n end")()
    bxor   = load("return function(a, b) return a ~ b end")()
  end
end

local STATE_RUNNING            = 0
local STATE_RUNNABLE           = 1
local STATE_SUSPENDED          = 2
local STATE_HELD               = 4
local STATE_SUSPENDED_RUNNABLE = 3
local STATE_NOT_HELD           = bnot(STATE_HELD)

local EXPECTED_QUEUE_COUNT = 232625
local EXPECTED_HOLD_COUNT  = 93050

local queueCount  = 0
local holdCount   = 0
local blocks      = {}
local list_head   = nil
local currentTcb  = nil
local currentId   = nil

local sched_release, sched_hold_current, sched_suspend_current, sched_queue
local task_run

-- ---- Packet ----
local function pkt_new(link, id, kind)
  return { link = link, id = id, kind = kind, a1 = 0, a2 = { 0, 0, 0, 0 } }
end

local function pkt_addto(packet, queue)
  packet.link = nil
  if queue == nil then return packet end
  local next = queue
  local peek = next.link
  while peek ~= nil do
    next = peek
    peek = next.link
  end
  next.link = packet
  return queue
end

-- ---- TCB ----
local function tcb_new(link, id, priority, queue, task)
  local state
  if queue == nil then state = STATE_SUSPENDED
  else state = STATE_SUSPENDED_RUNNABLE end
  return {
    link = link, id = id, priority = priority,
    queue = queue, task = task, state = state,
  }
end

local function tcb_held_or_suspended(t)
  return band(t.state, STATE_HELD) ~= 0 or t.state == STATE_SUSPENDED
end

local function tcb_check_priority_add(t, task, packet)
  if t.queue == nil then
    t.queue = packet
    t.state = bor(t.state, STATE_RUNNABLE)
    if t.priority > task.priority then return t end
  else
    t.queue = pkt_addto(packet, t.queue)
  end
  return task
end

local function tcb_run(t)
  local packet
  if t.state == STATE_SUSPENDED_RUNNABLE then
    packet = t.queue
    t.queue = packet.link
    if t.queue == nil then t.state = STATE_RUNNING
    else t.state = STATE_RUNNABLE end
  else
    packet = nil
  end
  return task_run(t.task, packet)
end

-- ---- Scheduler ops ----
local function add_task(id, priority, queue, task)
  local tcb = tcb_new(list_head, id, priority, queue, task)
  currentTcb = tcb
  list_head = tcb
  blocks[id] = tcb
end

local function add_running_task(id, priority, queue, task)
  add_task(id, priority, queue, task)
  currentTcb.state = STATE_RUNNING
end

sched_release = function(id)
  local tcb = blocks[id]
  if tcb == nil then return tcb end
  tcb.state = band(tcb.state, STATE_NOT_HELD)
  if tcb.priority > currentTcb.priority then return tcb end
  return currentTcb
end

sched_hold_current = function()
  holdCount = holdCount + 1
  currentTcb.state = bor(currentTcb.state, STATE_HELD)
  return currentTcb.link
end

sched_suspend_current = function()
  currentTcb.state = bor(currentTcb.state, STATE_SUSPENDED)
  return currentTcb
end

sched_queue = function(packet)
  local t = blocks[packet.id]
  if t == nil then return t end
  queueCount = queueCount + 1
  packet.link = nil
  packet.id = currentId
  return tcb_check_priority_add(t, currentTcb, packet)
end

-- ---- Tasks ----
local function idle_run(t, packet)
  t.count = t.count - 1
  if t.count == 0 then return sched_hold_current() end
  if band(t.v1, 1) == 0 then
    t.v1 = rshift(t.v1, 1)
    return sched_release(ID_DEVICE_A)
  else
    t.v1 = bxor(rshift(t.v1, 1), 0xD008)
    return sched_release(ID_DEVICE_B)
  end
end

local function device_run(t, packet)
  if packet == nil then
    if t.v1 == nil then return sched_suspend_current() end
    local v = t.v1
    t.v1 = nil
    return sched_queue(v)
  else
    t.v1 = packet
    return sched_hold_current()
  end
end

local function worker_run(t, packet)
  if packet == nil then return sched_suspend_current() end
  if t.v1 == ID_HANDLER_A then t.v1 = ID_HANDLER_B
  else t.v1 = ID_HANDLER_A end
  packet.id = t.v1
  packet.a1 = 0
  for i = 0, DATA_SIZE - 1 do
    t.v2 = t.v2 + 1
    if t.v2 > 26 then t.v2 = 1 end
    packet.a2[i + 1] = t.v2
  end
  return sched_queue(packet)
end

local function handler_run(t, packet)
  if packet ~= nil then
    if packet.kind == KIND_WORK then t.v1 = pkt_addto(packet, t.v1)
    else t.v2 = pkt_addto(packet, t.v2) end
  end
  if t.v1 ~= nil then
    local count = t.v1.a1
    if count < DATA_SIZE then
      if t.v2 ~= nil then
        local v = t.v2
        t.v2 = t.v2.link
        v.a1 = t.v1.a2[count + 1]
        t.v1.a1 = count + 1
        return sched_queue(v)
      end
    else
      local v = t.v1
      t.v1 = t.v1.link
      return sched_queue(v)
    end
  end
  return sched_suspend_current()
end

task_run = function(task, packet)
  return task.run(task, packet)
end

-- ---- Build helpers ----
local function add_idle_task(id, priority, queue, count)
  add_running_task(id, priority, queue,
    { run = idle_run, v1 = 1, count = count })
end
local function add_worker_task(id, priority, queue)
  add_task(id, priority, queue,
    { run = worker_run, v1 = ID_HANDLER_A, v2 = 0 })
end
local function add_handler_task(id, priority, queue)
  add_task(id, priority, queue,
    { run = handler_run, v1 = nil, v2 = nil })
end
local function add_device_task(id, priority, queue)
  add_task(id, priority, queue,
    { run = device_run, v1 = nil })
end

local function schedule()
  currentTcb = list_head
  while currentTcb ~= nil do
    if tcb_held_or_suspended(currentTcb) then
      currentTcb = currentTcb.link
    else
      currentId = currentTcb.id
      currentTcb = tcb_run(currentTcb)
    end
  end
end

-- ---- Run ----
add_idle_task(ID_IDLE, 0, nil, COUNT)

local q
q = pkt_new(nil, ID_WORKER, KIND_WORK)
q = pkt_new(q,   ID_WORKER, KIND_WORK)
add_worker_task(ID_WORKER, 1000, q)

q = pkt_new(nil, ID_DEVICE_A, KIND_DEVICE)
q = pkt_new(q,   ID_DEVICE_A, KIND_DEVICE)
q = pkt_new(q,   ID_DEVICE_A, KIND_DEVICE)
add_handler_task(ID_HANDLER_A, 2000, q)

q = pkt_new(nil, ID_DEVICE_B, KIND_DEVICE)
q = pkt_new(q,   ID_DEVICE_B, KIND_DEVICE)
q = pkt_new(q,   ID_DEVICE_B, KIND_DEVICE)
add_handler_task(ID_HANDLER_B, 3000, q)

add_device_task(ID_DEVICE_A, 4000, nil)
add_device_task(ID_DEVICE_B, 5000, nil)

schedule()

if queueCount ~= EXPECTED_QUEUE_COUNT or holdCount ~= EXPECTED_HOLD_COUNT then
  error("bad: queueCount=" .. queueCount .. " holdCount=" .. holdCount)
end

print(queueCount + holdCount)
